Enable noexcept for special member functions

pfent opened the pull request:
The hopscotch_hash container already was nothrow move constructible, but was missing the conditional noexcept for the default constructor, swap, and move assignment.

pfent committed:
Enable noexcept for special member functions

The hopscotch_hash container already was nothrow move constructible,
but was missing the conditional noexcept for the default constructor,
swap, and move assignment.

Tessil commented:
Thanks for the PR.

It looks good but the default constructor could still throw even with `DEFAULT_INIT_BUCKETS_SIZE == 0` as the `Hash`, `KeyEqual` and `GrowthPolicy` default constructors could potentially throw. We would need to add some check on these.

pfent committed:
Consider template args for default constructor noexceptness

This adds a helper to detect the default growth policies that are
noexcept when default initialized with 0.

pfent commented:
Thanks for the review, you raised a valid point:
The default growth policies are not trivially noexcept, so I added a small helper to help detect that the growth policies are actually noexcept when initialized with zero.

Tessil commented:
Sorry for the delay. 

The change looks good now. 

Just a minor question, could we create a default `hopscotch_hash` constructor with the `noexcept` checks and then for the default constructors of `(b)hopscotch_map/set` check the `noexcept` of the `hopscotch_hash` constructor (and call it)? Mainly to avoid all the duplication of the checks.

pfent committed:
Add a conditionally noexcept hopscotch_hash default constructor

This avoids duplication of is_nothrow_default_constructible checks.

pfent commented:
Good point. I moved the checks to `hopscotch_hash`, which makes toe overall diff a bit smaller.

Tessil commented:
Thanks, looks good.

Currently the library is compatible with C++11 (it was so when designed and kept like that) but `std::is_nothrow_swappable` requires C++17.

I'll move the library and tests to C++17 over the weekend as there's no real reason to keep C++11 compatibility and merge the change then.

Tessil commented:
I merged the changes. Note that I had to disable the `test_noexcept` test on MSVC as `std::list` is not `std::is_nothrow_move_constructible` in their implementation. Generally this overflow list should be empty (their need to be > 62 collisions on the same bucket for it to fill), but it's still blocking `noexcept` on MSVC. I have to check if there's eventually an alternative.
